### ISSUE TITLE
Cobaya run issues

### ISSUE THREAD

**Issue by nnedelko** [Lbalkenhol/candl#5]
Hi! I'm trying to do a cobaya MCMC run with the SPT-3G TTTEEE likelihood. The chains aren't failing, but I'm getting quite a few `candl/transformations/common.py:1340: RuntimeWarning: invalid value encountered in sqrt
  CIB_x_tSZ = jnp.sqrt(CIB_nu_1 * tSZ_nu_2) + jnp.sqrt(CIB_nu_2 * tSZ_nu_1)` warnings.

I'm also unsure whether or not I need to directly pass the SPT nuisance parameters in the cobaya .yaml input file. The documentation [here](https://candl.readthedocs.io/en/latest/interface/interface_sampler.html) seemingly says 'no', but in the `traditional_tutorial` notebook the nuisance parameters are directly added to the "params" block, is that being done automatically when invoking cobaya from console?

UPD: ok I've done some testing
I do need to list the nuisance parameters, makes sense
I did find a new odd behaviour: if I clear the internal priors and instead directly list them in the cobaya file, the run starts normally (aside from the warinings above), but if I remove the priors from the cobaya file (keeping all the other fields for the nuisance parameters and tau_reio the same) and set the `clear_internal_priors` flag to `false`, the run fails to start with a `[model] *ERROR* Could not find random point giving finite posterior after 200 tries`.

**Comment by nnedelko** [comment]
Additionally, I've found that the
```
likelihood:
    candl.interface.CandlCobayaLikelihood: 
      ...
```
likelihood statement from the "Samplers and minimizers" page doesn't work, it should be 
```
likelihood:
  candl_like:
    additional_args: {}
    external: !!python/name:candl.interface.CandlCobayaLikelihood ''
    ...
```
which is what I get when converting the invocation from the tutorial notebook to yaml

**Comment by nnedelko** [comment]
Hmm. I have candl 1.7.0 isntalled from pip, so it should be up to date.
What's odd to me about the final point error is that I *am* giving cobaya both the `proposal` and the `ref` fields for all nuisance parameters, so to my understanding the starting point in that part of the parameter space should be fixed:

```
likelihood:
   candl_like:
    additional_args: {}
    lensing: false
    clear_internal_priors: false
    data_set_file: /lustre/home/user/n/nnedelko/.local/lib/python3.10/site-packages/candl/data/SPT3G_2018_TTTEEE_v0/SPT3G_2018_TTTEEE.yaml
    external: !!python/name:candl.interface.CandlCobayaLikelihood ''
    feedback: true
    wrapper: null
params:  
  logA:
    prior:
      min: 1.61
      max: 3.91
    ref:
      dist: norm
      loc: 3.05
      scale: 0.001
    proposal: 0.001
    latex: \log(10^{10} A_\mathrm{s})
    drop: true
  A_s:
    value: 'lambda logA: 1e-10*np.exp(logA)'
    latex: A_\mathrm{s}
  n_s:
    prior:
      min: 0.8
      max: 1.2
    ref:
      dist: norm
      loc: 0.965
      scale: 0.004
    proposal: 0.002
    latex: n_\mathrm{s}
  H0:
    prior:
      min: 20
      max: 100
    ref:
      dist: norm
      loc: 69
      scale: 1
    proposal: 1
    latex: H_0
  omega_b:
    prior:
      min: 0.005
      max: 0.1
    ref:
      dist: norm
      loc: 0.0224
      scale: 0.0001
    proposal: 0.0001
    latex: \Omega_\mathrm{b} h^2
  omega_cdm:
    prior:
      min: 0.001
      max: 0.99
    ref:
      dist: norm
      loc: 0.12
      scale: 0.001
    proposal: 0.0005
    latex: \Omega_\mathrm{c} h^2
  Omega_m:
    latex: \Omega_\mathrm{m}
  omegamh2:
    derived: 'lambda Omega_m, H0: Omega_m*(H0/100)**2'
    latex: \Omega_\mathrm{m} h^2
  m_ncdm:
    value: 0.06
    renames: mnu
  Omega_Lambda:
    latex: \Omega_\Lambda
  YHe:
    latex: Y_\mathrm{P}
  tau_reio:
    ref:
      dist: norm
      loc: 0.054
      scale: 0.01
    proposal: 0.005
    latex: \tau_\mathrm{reio}
  z_reio:
    latex: z_\mathrm{re}
  sigma8:
    latex: \sigma_8
  s8h5:
    derived: 'lambda sigma8, H0: sigma8*(H0*1e-2)**(-0.5)'
    latex: \sigma_8/h^{0.5}
  s8omegamp5:
    derived: 'lambda sigma8, Omega_m: sigma8*Omega_m**0.5'
    latex: \sigma_8 \Omega_\mathrm{m}^{0.5}
  s8omegamp25:
    derived: 'lambda sigma8, Omega_m: sigma8*Omega_m**0.25'
    latex: \sigma_8 \Omega_\mathrm{m}^{0.25}
  A:
    derived: 'lambda A_s: 1e9*A_s'
    latex: 10^9 A_\mathrm{s}
  clamp:
    derived: 'lambda A_s, tau_reio: 1e9*A_s*np.exp(-2*tau_reio)'
    latex: 10^9 A_\mathrm{s} e^{-2\tau}
  age:
    latex: '{\rm{Age}}/\mathrm{Gyr}'
  rs_drag:
    latex: r_\mathrm{drag}
  
  EE_Poisson_150x150:
    latex: EE_Poisson_150x150
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - &id001 !!python/object/apply:numpy.dtype
      args:
      - f8
      - false
      - true
      state: !!python/tuple
      - 3
      - <
      - null
      - null
      - null
      - -1
      - -1
      - 0
    - !!binary |
      8UqS5/o+bD8=
    ref: 0.011495
  EE_Poisson_150x220:
    latex: EE_Poisson_150x220
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      wK4mT1lNdz8=
    ref: 0.018962
  EE_Poisson_220x220:
    latex: EE_Poisson_220x220
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      b9dLUwQ4jT8=
    ref: 0.047557
  EE_Poisson_90x150:
    latex: EE_Poisson_90x150
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      3ncMj/0sdj8=
    ref: 0.018048
  EE_Poisson_90x220:
    latex: EE_Poisson_90x220
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      QndJnBVRcz8=
    ref: 0.015719
  EE_Poisson_90x90:
    latex: EE_Poisson_90x90
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      maCGb2HdiD8=
    ref: 0.040469
  EE_PolGalDust_Alpha:
    latex: EE_PolGalDust_Alpha
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      exSuR+F6pD8=
    ref: -2.42
  EE_PolGalDust_Amp:
    latex: EE_PolGalDust_Amp
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      ukkMAiuHlj8=
    ref: 0.05
  EE_PolGalDust_Beta:
    latex: EE_PolGalDust_Beta
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      exSuR+F6pD8=
    ref: 1.51
  Ecal150:
    latex: Ecal150
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      xTBFM0K9gD8=
    ref: 1.0
  Ecal220:
    latex: Ecal220
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      l8BMQOpnkD8=
    ref: 1.0
  Ecal90:
    latex: Ecal90
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      py7Bc/HbgT8=
    ref: 1.0
  Kappa:
    latex: Kappa
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      kst/SL99PT8=
    ref: 0.0
  TE_PolGalDust_Alpha:
    latex: TE_PolGalDust_Alpha
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      exSuR+F6pD8=
    ref: -2.42
  TE_PolGalDust_Amp:
    latex: TE_PolGalDust_Amp
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      6SYxCKwcqj8=
    ref: 0.12
  TE_PolGalDust_Beta:
    latex: TE_PolGalDust_Beta
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      exSuR+F6pD8=
    ref: 1.51
  TT_CIBClustering_Amp:
    latex: TT_CIBClustering_Amp
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      IEHxY8xd/T8=
    ref: 3.2263
  TT_CIBClustering_Beta:
    latex: TT_CIBClustering_Beta
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      DXGsi9to2D8=
    ref: 2.2642
  TT_GalCirrus_Alpha:
    latex: TT_GalCirrus_Alpha
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      mpmZmZmZqT8=
    ref: -2.53
  TT_GalCirrus_Amp:
    latex: TT_GalCirrus_Amp
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      uB6F61G43j8=
    ref: 1.88
  TT_GalCirrus_Beta:
    latex: TT_GalCirrus_Beta
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      exSuR+F6lD8=
    ref: 1.48
  TT_Poisson_150x150:
    latex: TT_Poisson_150x150
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      ukkMAiuHEEA=
    ref: 15.3455
  TT_Poisson_150x220:
    latex: TT_Poisson_150x220
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      hetRuB7FEEA=
    ref: 28.3573
  TT_Poisson_220x220:
    latex: TT_Poisson_220x220
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      CyQofoy5LUA=
    ref: 75.9719
  TT_Poisson_90x150:
    latex: TT_Poisson_90x150
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      Qxzr4jZaHEA=
    ref: 22.4417
  TT_Poisson_90x220:
    latex: TT_Poisson_90x220
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      i2zn+6mxF0A=
    ref: 20.7004
  TT_Poisson_90x90:
    latex: TT_Poisson_90x90
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      /Knx0k3iIkA=
    ref: 51.3204
  TT_kSZ_Amp:
    latex: TT_kSZ_Amp
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      +n5qvHSTEkA=
    ref: 3.7287
  TT_tSZ_Amp:
    latex: TT_tSZ_Amp
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      F7fRAN4CA0A=
    ref: 3.2279
  TT_tSZ_CIB_Corr_Amp:
    latex: TT_tSZ_CIB_Corr_Amp
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      T6+UZYhj1T8=
    ref: 0.1801
  Tcal150:
    latex: Tcal150
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      eGpIMeb6dj8=
    ref: 1.0
  Tcal220:
    latex: Tcal220
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      4RAO2PCWfj8=
    ref: 1.0
  Tcal90:
    latex: Tcal90
    proposal: !!python/object/apply:numpy.core.multiarray.scalar
    - *id001
    - !!binary |
      QxyhCvTQdj8=
    ref: 1.0
```

I'll try replacing the `ref` values with the prior distributions.

**Comment by nnedelko** [comment]
Oh, by the way, if I'm now understanding the use of priors correctly, in block [9] of traditional_tutorial.ipynb you are double counting the priors because the `clear_internal_priors` flag is set to `false` but you are then directly imposing the priors inside cobaya. Of course it doesn't make a huge difference for a sample run with just 100 accepted steps, but still.

**Comment by Lbalkenhol** [comment]
Glad the lite variant is proving useful.

I don’t think the priors are correctly applied in the snippet you sent. There are 2 types of priors at play here, both of which we want:

(1) Gaussian priors
(2) Uniform “box” priors

candl internally only applies priors of type (1) (the reason being that box priors are not differentiable at the boundaries). However priors of type (2) are still needed: there are some parameters which are not very well constrained by data+Gaussian prior together and may very well intersect boundaries that we think should exist. The tSZ-CIB correlation amplitude for example, which as a correlation coefficient should be between -1 and 1, but is difficult to constrain in this kind of analysis.

As before, there are two solutions here:

(1) Do all priors in cobaya. Set `clear_internal_priors = True` for candl. Add gaussian priors all on nuisance parameters and specify the parameter limits too. This is more transparent, but you need to pay special attention to the multi-dimensional prior on calibration parameters when doing so.

This would look like this for example (in the cobaya yaml file):
```
params:
  EE_Poisson_90x90:
    prior:
      max: 5.0
      min: 0.0
    ref: 0.040469
prior:
  gaussian_EE_Poisson_90x90: 'lambda EE_Poisson_90x90: stats.norm.logpdf(EE_Poisson_90x90, 0.040469, 0.012141)
```

(2) Keep `clear_internal_priors=False` for candl, but enforce parameter limits in Cobaya. You then just set prior: max/min in the cobaya block and don’t worry about anything else. It’s less clean maybe, as box priors are now applied by Cobaya and Gaussian ones by candl, but it’s less work.

With this in mind the tutorial is fine. Box priors are applied in Cobaya, Gaussian ones in candl. Both need to be present.

**Comment by nnedelko** [comment]
Ah I see, thanks for the explanation. I'm new to working with cobaya after montepython so I'm used to everything other than box priors being separated into the likelihood files. Yeah in that case the invalid sqrt is almost certainly caused by the missing box priors. I'll close the issue now and come back if providing all priors won't work. Thanks again.

**Comment by nnedelko** [comment]
This

> there are some parameters which are not very well constrained by data+Gaussian prior together and may very well intersect boundaries that we think should exist. The tSZ-CIB correlation amplitude for example, which as a correlation coefficient should be between -1 and 1, but is difficult to constrain in this kind of analysis.


was indeed the issue. More specifically, the method I was using to get the box priors, 
```
cobaya_info["params"][par_name] = {'prior':
                                               {'min': prior_central_val-10*prior_width,
                                                'max': prior_central_val+10*prior_width},
                                               'ref': prior_central_val,
                                               'proposal': prior_width,
                                               'latex': par_name}

```
creates a simple symmetric 20*prior_width box and so doesn't check if this accidentally allows for e.g. invalid amplitude values. I've manually replaced these box priors with the ones from the CosmoMC likelihood and that fixed the issue.

**Comment by nnedelko** [comment]
> you need to pay special attention to the multi-dimensional prior on calibration parameters when doing so

Am I correctly understanding that in Cobaya the way to do this is putting the multivatiariate normal function with into a separate file and then calling it as an external likelihood?

I would like to suggest adding an option to *only remove the prior on tau* (for example, I'm running a combination of SPT-3G and Planck data and so I want to use the lowl-EE Planck likelihood instead of a gaussian prior on tau, it would be very nice to be able to do that without having to fully clear all internal SPT priors and then redefining them all manually)

**Comment by Lbalkenhol** [comment]
I think you should be able to extend this 1-d syntax to higher dimensions without having to resort to having an external likelihood:
```
prior:
  gaussian_EE_Poisson_90x90: 'lambda EE_Poisson_90x90: stats.norm.logpdf(EE_Poisson_90x90, 0.040469, 0.012141)
```
without testing, something like 
```
prior:
  2d_prior: 'lambda par1, par2: stats.multivariate_normal.logpdf(np.array([par1, par2]), 0, np.eye(2))
```
might work (using [this scipy function](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.multivariate_normal.html)).

> I would like to suggest adding an option to only remove the prior on tau

I think that's a good idea, thanks for the suggestion. I'm hoping to find some time soon to push a few improvements. In the mean-time you could consider editing the candl data set .yaml files (or making copies of them) and removing the tau prior. Using the [variants syntax](https://candl.readthedocs.io/en/latest/like/like_data_structure.html#index-files) you could label your own "no_tau_prior" variants.

**Comment by Lbalkenhol** [comment]
The latest version I just published has an option to supply `clear_specific_priors` to the cobaya wrapper. To only remove the tau prior, but keep all other internal priors use:
```
clear_internal_priors = False
clear_specific_priors = "tau"
```
More info is in the [docs](https://candl.readthedocs.io/en/stable/interface/interface_sampler.html).

Regarding the issue of pointing to `candl.interface.CandlCobayaLikelihood` in the yaml file, I'm having trouble reproducing this on my end, so I've added the workaround to the docs, thanks.
